File: examples/openai_agents.py
"""
OpenAI Agents SDK — traced by OrchestraAI.

Auto-instruments the OpenAI Python SDK to capture every chat completion call
with token usage, model name, latency, and I/O previews — zero code changes
to your agent logic.

Prerequisites:
    pip install -e sdks/python
    pip install openai

Run:
    export ORCHESTRA_API_KEY=oai_...
    export OPENAI_API_KEY=sk-...
    python examples/openai_agents.py
"""

import os
from openai import OpenAI
from orchestra_ai import OrchestraAI
from orchestra_ai.integrations import openai_agents_tracer

# ── Setup ──────────────────────────────────────────────────────
oa = OrchestraAI(
    api_key=os.environ["ORCHESTRA_API_KEY"],
    base_url=os.getenv("ORCHESTRA_BASE_URL", "http://localhost:3001"),
)

# Auto-instrument OpenAI SDK — patches Completions.create globally
openai_agents_tracer.auto_instrument(oa)

client = OpenAI()  # Uses OPENAI_API_KEY from env

# ── Run agent logic (every LLM call is auto-traced) ───────────
print("=== OpenAI SDK with OrchestraAI Auto-Instrumentation ===\n")

# Call 1: Planning
print("1. Planning step...")
plan = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {"role": "system", "content": "You are a research assistant. Plan your approach."},
        {"role": "user", "content": "I need a summary of recent advances in AI agent safety."},
    ],
    max_tokens=200,
)
print(f"   Plan: {plan.choices[0].message.content[:150]}...")

# Call 2: Research
print("\n2. Research step...")
research = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {"role": "system", "content": "You are a research analyst. Be thorough but concise."},
        {"role": "user", "content": "What are the key approaches to AI agent safety in 2024-2025?"},
    ],
    max_tokens=400,
)
print(f"   Research: {research.choices[0].message.content[:150]}...")

# Call 3: Synthesis
print("\n3. Synthesis step...")
summary = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {"role": "system", "content": "Synthesize the following into a 3-sentence summary."},
        {"role": "user", "content": research.choices[0].message.content or ""},
    ],
    max_tokens=150,
)
print(f"   Summary: {summary.choices[0].message.content}")

# Show total usage
total_in = sum(r.usage.prompt_tokens for r in [plan, research, summary] if r.usage)
total_out = sum(r.usage.completion_tokens for r in [plan, research, summary] if r.usage)
print(f"\nTotal tokens: {total_in} in / {total_out} out")

print("\nAll 3 LLM calls were auto-traced to OrchestraAI!")
print("Check traces at http://localhost:3000/dashboard/traces")

# Clean up
openai_agents_tracer.remove_instrumentation()
oa.close()

File: examples/openai_agents.ts
/**
 * OpenAI SDK — traced by OrchestraAI (TypeScript).
 *
 * Uses the SDK's auto token extraction — just pass the response object.
 * No monkey-patching; explicit tracing with zero boilerplate.
 *
 * Prerequisites:
 *   npm link ./sdks/typescript
 *   npm install openai
 *
 * Run:
 *   export ORCHESTRA_API_KEY=oai_...
 *   export OPENAI_API_KEY=sk-...
 *   npx tsx examples/openai_agents.ts
 */

import { OrchestraAI } from '../sdks/typescript/src';
import OpenAI from 'openai';

const oa = new OrchestraAI({
  apiKey: process.env.ORCHESTRA_API_KEY!,
  baseUrl: process.env.ORCHESTRA_BASE_URL || 'http://localhost:3001',
});

const openai = new OpenAI();

async function main() {
  console.log('=== OpenAI SDK with OrchestraAI Tracing (TypeScript) ===\n');

  await oa.trace('openai-research-agent', async (trace) => {
    // Step 1: Planning
    console.log('1. Planning...');
    const plan = await openai.chat.completions.create({
      model: 'gpt-4o-mini',
      messages: [
        { role: 'system', content: 'You are a research assistant. Plan your approach.' },
        { role: 'user', content: 'I need a summary of AI agent safety approaches.' },
      ],
      max_tokens: 200,
    });
    await trace.llmCall({
      response: plan,
      inputPreview: 'Plan: summary of AI agent safety approaches',
      outputPreview: plan.choices[0].message.content ?? '',
    });
    console.log(`   Plan: ${plan.choices[0].message.content?.slice(0, 120)}...`);

    // Step 2: Tool call (simulated)
    console.log('2. Searching docs...');
    const toolSpan = trace.toolCall({
      toolName: 'arxiv_search',
      toolInput: { query: 'AI agent safety 2024' },
    });
    toolSpan.setData({ toolOutput: 'Found 5 papers on AI agent safety' });
    toolSpan.end();

    // Step 3: Synthesis
    console.log('3. Synthesizing...');
    const summary = await openai.chat.completions.create({
      model: 'gpt-4o-mini',
      messages: [
        { role: 'system', content: 'Summarize in 3 sentences.' },
        { role: 'user', content: plan.choices[0].message.content || '' },
      ],
      max_tokens: 150,
    });
    await trace.llmCall({
      response: summary,
      inputPreview: 'Synthesize the plan into 3 sentences',
      outputPreview: summary.choices[0].message.content ?? '',
    });
    console.log(`   Summary: ${summary.choices[0].message.content}`);
  }, { sessionId: 'research-session-001' });

  console.log('\nAll calls traced! Check http://localhost:3000/dashboard/traces');
}

main().catch(console.error);
